### ISSUE TITLE
[vault-secrets-webhook] namespaced access to secrets

### ISSUE THREAD

**Issue by philband** [banzaicloud/bank-vaults#929]
**Is your feature request related to a problem? Please describe.**
I'm currently trying out the vault-secrets-webhook on top of an existing Vault deployment.
I'm currently using a vault-template sidecar along with existing deployments to load secrets and trying to figure out how to migrate to the webhook.

For authentication to Vault I'm using a dedicated ServiceAccount in each namespace. Therefore the vault policies and the roles used to authenticate are also namespaced to the specific deployments running in the cluster. Apps can only access the secret their individual ServiceAccount from their namespace has access to.

As far as I understand the vault-secrets-webhook, access to all Vault data is authenticated with a single ServiceAccount, residing in the namespace of the webhook.

**Describe the solution you'd like**
From a security perspective, I would like to have an option to let the webhook use a ServiceAccount in the namespace of the to-be-mutated resource. Otherwise, it would be possible for any resource in any namespace to query for any secret that the webhook has access to.

**Describe alternatives you've considered**
It is quite possible, that this is already possible with the webhook and I didn't figure out. If not I would propose to allow the webhook to be configured to use a ServiceAccount in the namespace of the resource requesting a secret, possibly also auto-creating this service account

**Comment by bonifaido** [comment]
Hi @philband, I suggest to read and the blogs: https://banzaicloud.com/docs/bank-vaults/mutating-webhook/ but in a nutshell: when injecting into Pods entities always the ServiceAccount of that Pod is used to access Vault, let it be vault-env or the sidecar version.
For injecting into Kubernetes Secrets, the ServiceAccount of the webhook is used.

**Comment by philband** [comment]
Hi @bonifaido, thank you first of all for the quick response. 
I now realize, that my question was not as clear as I intended. How this works for adding secrets to deployments etc. is clear to me. 

The question/request was intended to ask, whether it is possible to use a custom ServiceAccount for mutating Secrets and/or CRDs?

**Comment by bonifaido** [comment]
Yeah, that is not possible currently. And I wonder how could we implement that.

**Comment by philband** [comment]
After reviewing the webhook codebase, I think I might have an idea on how to implement this feature.

I would propose to add an optional annotation `vault.security.banzaicloud.io/vault-serviceaccount`, which would hold the name of a service account in the namespace of the `Secret`, `ConfigMap` or `CustomResource` to be mutated. 

If this annotation is present, the webhook shall query the k8s API for a `ServiceAccount` with that name in the namespace of the resource requesting mutation. From that, it should fetch the `jwt` associated with the `ServiceAccount`. 

With this `jwt`, the webhook can authenticate against vault, as if the request came from the `ServiceAccount` specified in the annotation.

This `ServiceAccount` would have to be handed down to the vault client wrapper, replacing the default `ServiceAccount` token injected into the webhook-pod:
https://github.com/banzaicloud/bank-vaults/blob/7fadfb8416b0d1cd3ac5db5aba5d94e56ac50b8f/pkg/sdk/vault/client.go#L274

If I'm not mistaken, this would also require adjustments in the helm chart for the webhook (RBAC), to allow the webhook to query `ServiceAccount` resources.

Should you be agreeable to this proposal, I will try to make a PR for it.

**Comment by bonifaido** [comment]
The technical part of the proposal sounds right to me. But giving the webhook the right to query and read service account tokens sounds a bit a broad scope of privileges that concerns me. If the webhook as all the rights to read service accounts people won't install it, if we can make this more scoped to only a list of service accounts that might be okay.

**Comment by philband** [comment]
Current state, according to the helm chart of the webhook:
https://github.com/banzaicloud/bank-vaults/blob/7fadfb8416b0d1cd3ac5db5aba5d94e56ac50b8f/charts/vault-secrets-webhook/templates/webhook-rbac.yaml#L8-L27
The webhook can already view and update `Secrets` and `ConfigMaps` across all namespaces.

So here is another idea: 
```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: vault-webhook-mutator
rules:
  - apiGroups:
      - ""
    resources:
      - serviceaccounts
    verbs:
      - "create"
      - "get"
      - "update"
    resourceNames:
      - "vault-webhook-mutator"
```
We give the webhook permissions to create, get and update a `ServiceAccount` with a well-known name (e.g. `vault-webhook-mutator`), that is just used for the purposes of the mutation.
The `jwt` attached to the secret referenced by the `ServiceAccount` can already be read.

One thing to consider in this idea: Should the webhook auto-create the `ServiceAccount` if it is missing? Or is this a task to be done by the user?

**Comment by bonifaido** [comment]
Sounds good to me! 👍 

I think we should leave the SA creation to the users IMHO.

**Comment by ron1** [comment]
@philband Have you made any progress on a PR for this feature? The lack of namespace isolation in the current solution makes it unusable for us.

**Comment by bonifaido** [comment]
@ron1 would be the configurable auto-creation of the ServiceAccount enough for you?

> One thing to consider in this idea: Should the webhook auto-create the ServiceAccount if it is missing? Or is this a task to be done by the user?

**Comment by ron1** [comment]
@bonifaido Yes, configurable ServiceAccount creation would work for us. The GoDaddy Kubernetes External Secrets controller uses a simpler strategy to solve this problem which would also work for us. This controller introduces a namespace annotation to enforce naming conventions for backend keys as described here: https://github.com/godaddy/kubernetes-external-secrets#enforcing-naming-conventions-for-backend-keys. If managing specific, per-namespace service accounts is too cumbersome or introduces too much overhead, using naming conventions to control backend key access would be sufficient for us. WDYT?

**Comment by bonifaido** [comment]
I like this namespace annotation idea TBH. But this needs to be implemented in the webhook I guess.

**Comment by ron1** [comment]
This seems like a reasonably straight-forward validation that could be performed by a webhook against the Secret/ConfigMap upon creation or edit. Do you agree?

I suspect this approach would also solve issue https://github.com/banzaicloud/bank-vaults/issues/964 in a more streamlined manner than creating individual webhooks to control specific namespaces.

**Comment by bonifaido** [comment]
Great, I absolutely agree! Are you about to give it a stab @ron1?

**Comment by ron1** [comment]
@bonifaido When ConfigMaps and Secrets are parsed using envFrom in a Pod, is the serviceAccount of the Pod or the WebHook used when accessing Vault.

**Comment by ron1** [comment]
@bonifaido @philband After looking at the following code in the GoDaddy Kubernetes External Secrets controller that enforces the namespace isolation: https://github.com/godaddy/kubernetes-external-secrets/blob/71901209cf76491948072f675e1732cb7efdffbf/lib/poller.js#L204, I think the @philband idea of enforcing namespace isolation using standard Vault constructs rather than webhook-applied regex tests is preferred.

If a user wants Secrets/ConfigMaps mutated in a namespace, I think it is reasonable to require that they create a vault-webhook-mutator SA in that namespace and that they define a Vault Role and accompanying Policy for that SA. Attempting to create or update a Secret/ConfigMap that has a vault url in a namespace without an appropriately entitled vault-webhook-mutator SA would fail validation. Does that seem reasonable?

**Comment by ron1** [comment]
@bonifaido Note that the below comment on the vault-role annotation in doc section https://banzaicloud.com/docs/bank-vaults/mutating-webhook/#example leads me to believe that the vault-webhook-mutator already (by default) uses the `default` SA in the namespace to control access to vault for Secrets and ConfigMaps. Based on your comments above, I presume I am misinterpreting that comment included below. Or does the comment only apply to the use of Secrets and ConfigMaps referenced in Pods via envFrom?

```
      annotations:
        ...
        vault.security.banzaicloud.io/vault-role: "default" # optional, the default value is the name of the ServiceAccount the Pod runs in, in case of Secrets and ConfigMaps it is "default"
```

**Comment by bonifaido** [comment]
> @bonifaido When ConfigMaps and Secrets are parsed using envFrom in a Pod, is the serviceAccount of the Pod or the WebHook used when accessing Vault. It is not clear to me in this doc: https://banzaicloud.com/docs/bank-vaults/mutating-webhook/.

When there is runtime entity mutated, like a Pod which has a serviceAccount then always its SA  token is used to access Vault, even when using `envFrom`, since that is just a "syntactic sugar".

For mutating ConfigMaps/Secrets/Any Objects the webhook SA is used.

**Comment by alex-vmw** [comment]
@bonifaido @ron1 Do you guys have any ETA for when the new `vault.security.banzaicloud.io/vault-serviceaccount` annotation might be available to use? Thanks.

**Comment by bonifaido** [comment]
We don't have an ETA, for this, PRs are welcome! 🙂

**Comment by aelbarkani** [comment]
Would this feature be compatible with OIDC issuer discovery https://banzaicloud.com/blog/kubernetes-oidc/ ?

**Comment by aelbarkani** [comment]
@ron1 did you find any workaround for your use case ? The lack of namespace isolation makes it unusable for us too :(
